library/core/src/ptr/iter.rs

```diff
@@ -0,0 +1,513 @@
+use crate::fmt::{self, Debug};
+use crate::hash::{Hash, Hasher};
+use crate::iter::{FusedIterator, TrustedLen, TrustedRandomAccess, TrustedRandomAccessNoCoerce};
+use crate::mem;
+
+macro_rules! ptr_iter {
+    ($ty:ident, *$mutability:tt T) => {
+        /// Iterator with items generated by incrementing a raw pointer.
+        ///
+        /// # Safety
+        ///
+        /// Constructing an iterator, as well as dereferencing any pointers
+        /// obtained from traversing that iterator, require several invariants
+        /// to be upheld by the programmer or else the result is Undefined
+        /// Behavior.
+        #[doc = concat!("See [`", stringify!($ty), "::new`].")]
+        ///
+        /// Most importantly, remember that the pointers between the start and
+        /// end of a range "remember" the [allocated object] that they refer
+        /// into. Pointers resulting from pointer arithmetic must not be used to
+        /// read or write to any other allocated object.
+        ///
+        /// As a consequence, pointers from a traversal of this iterator are
+        /// only dereferenceable if the `start` and `end` with which the
+        /// iterator was constructed both point into the same allocated object.
+        /// Dereferencing a pointer obtained via iteration when this is not the
+        /// case is Undefined Behavior.
```

Comment:
I think this paragraph could use some tweaking.  `new` now requires the same allocated object (explicitly in safety text and indirectly via talking about being equivalent to looping via `offset(1)`), so I think the "same object" UB has already been hit by the iteration before the dereference comes into play now.

(Even `iterconst.next().unwrap()` has run an `offset(1)`, even if you only ever look at the first pointer.)